src/meta/processors/admin/BalancePlan.cpp

```diff
@@ -84,13 +84,19 @@ void BalancePlan::invoke() {
                         finished = true;
                         LOG(INFO) << "Balance " << id_ << " failed!";
                     }
+                    stopped = stopped_;
                 }
                 if (finished) {
                     CHECK_EQ(j, this->buckets_[i].size() - 1);
                     saveInStore(true);
                     onFinished_();
                 } else if (j + 1 < this->buckets_[i].size()) {
                     auto& task = this->tasks_[this->buckets_[i][j + 1]];
+                    if (tasks_[taskIndex].spaceId_ == task.spaceId_
```

Comment:
Is this why we sometimes recover, and still have failed tasks?

Reply:
The logic here is to avoid messing up peers if the previous task failed for the same part.